lightly_studio/src/lightly_studio/models/video.py

```diff
@@ -0,0 +1,150 @@
+"""This module defines the Video and VideoFrame model for the application."""
+
+from typing import TYPE_CHECKING, List, Optional
+from uuid import UUID
+
+from pydantic import BaseModel, ConfigDict
+from pydantic import Field as PydanticField
+from sqlalchemy.orm import Mapped
+from sqlmodel import Field, Relationship, SQLModel
+
+if TYPE_CHECKING:
+    from lightly_studio.models.annotation.annotation_base import (
+        AnnotationBaseTable,
+    )
+    from lightly_studio.models.sample import SampleTable, SampleView
+
+else:
+    AnnotationBaseTable = object
+    SampleTable = object
+
+
+# VideoFrame
+
+class VideoFrameBase(SQLModel):
+    """Base class for the VideoFrame model."""
+
+    """The frame number of the video frame."""
+    frame_number: int
+
+    """The timestamp of the video frame."""
+    frame_timestamp: int
+
+    """The video ID to which the video frame belongs."""
+    video_sample_id: UUID = Field(default=None, foreign_key="video.sample_id")
+
+
+class VideoFrameCreate(VideoFrameBase):
+    """VideoFrame class when inserting."""
+
+
+class VideoFrameTable(VideoFrameBase, table=True):
+    """This class defines the VideoFrame model."""
+
+    __tablename__ = "video_frame"
+    sample_id: UUID = Field(foreign_key="sample.sample_id", primary_key=True)
+    annotations: Mapped[List["AnnotationBaseTable"]] = Relationship(
+        back_populates="sample",
+    )
+
+    sample: Mapped["SampleTable"] = Relationship()
+    video: Mapped["VideoTable"] = Relationship(back_populates="frames")
+
+
+class VideoFrameView(SQLModel):
+    """VideoFrame class when retrieving."""
+
+    class VideoFrameViewVideo(SQLModel):
+        """Video view inside VideoFrame view."""
+        file_name: str
+        file_path_abs: str
+        sample_id: UUID
+        width: int
+        height: int
+        duration: float
+        fps: float
+
+    """The frame number of the video frame."""
+    frame_number: int
+    frame_timestamp: int
+    sample_id: UUID
+    video_sample_id: UUID
+
+    # Video metadata routed from parent video
+    video: VideoFrameViewVideo
+    sample: SampleView
+
+
+class VideoFrameViewsWithCount(BaseModel):
+    """Response model for counted video frames."""
+
+    model_config = ConfigDict(populate_by_name=True)
+
+    samples: List[VideoFrameView] = PydanticField(..., alias="data")
```

Comment:
This should be renamed to video_frames. The name `samples` is legacy, I did not manage to rename everything to `images`.